src/lib/encryption.ts

```diff
@@ -0,0 +1,77 @@
+import { Poseidon, Group, Field, Scalar } from '../snarky';
+import { PrivateKey, PublicKey } from './signature';
+
+export { encrypt, decrypt };
+
+class Sponge {
+  sponge: unknown;
+
+  constructor() {
+    this.sponge = (Poseidon as any).spongeCreate()();
+  }
+
+  absorb(x: Field) {
+    // console.log(this.sponge, x);
+    (Poseidon as any).spongeAbsorb(this.sponge, x);
+  }
+
+  squeeze() {
+    return (Poseidon as any).spongeSqueeze(this.sponge);
+  }
+}
+
+Poseidon.Sponge = Sponge;
+
+type CipherText = {
+  publicKey: Group;
+  cipherText: Field[];
+};
+
+function encrypt(message: Field[], otherPublicKey: PublicKey) {
+  // key exchange
+  let privateKey = Scalar.random();
```

Comment:
shouldn't you make sure that this is run as a witness thingy? I can't remember the exact API for that, but I'm wondering what `Scalar.random()` means in a circuit otherwise

Reply:
Yep the underlying implementation in snarky_js_bindings creates variables [here](https://github.com/MinaProtocol/mina/blob/0587489752ea9bf4334455fa6301313a8a188c35/src/lib/snarky_js_bindings/lib/snarky_js_bindings_lib.ml#L801) (actually an array of boolean vars representing the bits of a scalar, which seems pretty awful to me)